core/arch/arm/kernel/generic_entry_a32.S

```diff
@@ -46,19 +60,102 @@ FUNC _start , :
 	b	.	/* FIQ */
 END_FUNC _start
 
+/* Let platforms override this if needed */
+.weak plat_cpu_reset_early
+FUNC plat_cpu_reset_early , :
+	bx	lr
+END_FUNC plat_cpu_reset_early
+
+.weak plat_cpu_reset_late
+FUNC plat_cpu_reset_late , :
+	bx	lr
+END_FUNC plat_cpu_reset_late
+
+#ifdef CFG_BOOT_SYNC_CPU
+LOCAL_FUNC cpu_is_ready , :
+	lsl	r0, r0, #2
+	ldr	r1,=sem_cpu_sync
+	ldr	r2, =SEM_CPU_READY
+	str	r2, [r1, r0]
+	dsb
+	sev
+	bx	lr
+END_FUNC cpu_is_ready
+
+LOCAL_FUNC wait_primary , :
+	ldr	r0, =sem_cpu_sync
+	mov	r2, #SEM_CPU_READY
+	sev
+_wait_cpu0:
+	ldr	r1, [r0]
+	cmp	r1, r2
+	wfene
+	bne	_wait_cpu0
+	bx	lr
+END_FUNC wait_primary
+
+LOCAL_FUNC wait_secondary , :
+	ldr	r0, =sem_cpu_sync
+	mov	r3, #CFG_TEE_CORE_NB_CORE
+	mov	r2, #SEM_CPU_READY
+	sev
+_wait_next:
+	subs	r3, r3, #1
+	beq	_synced_cpun
+	add	r0, r0, #4
+_wait_cpun:
+	ldr	r1, [r0]
+	cmp	r1, r2
+	wfene
+	bne	_wait_cpun
+	b	_wait_next
+_synced_cpun:
+	bx	lr
+END_FUNC wait_secondary
+
+#else
+/*
+ * CFG_BOOT_SYNC_CPU is not defined
+ * So cpu synchronization functions are empty
+ */
+LOCAL_FUNC cpu_is_ready , :
+	bx	lr
+END_FUNC cpu_is_ready
+
+LOCAL_FUNC wait_primary , :
+	bx	lr
+END_FUNC wait_primary
+
+LOCAL_FUNC wait_secondary , :
+	bx	lr
+END_FUNC wait_secondary
+#endif
+
 LOCAL_FUNC reset , :
 	mov	r4, r0		/* Save pageable part address */
 	mov	r5, lr		/* Save ns-entry address */
 
+#ifdef CFG_TEE_GDB_BOOT
 	/*
-	 * Enable alignment checks and disable data and instruction cache.
+	 * boot preloaded linux: save linux boot args from GDB
+	 * all following routines, until reset_primary is called,
+	 * preserve R10/R11/R12
 	 */
+	mov	r10, r0
+	mov	r11, r1
+	mov	r12, r2
```

Comment:
r12 is a scratch register according to "Procedure Call Standard for the ARM® Architecture". A safer choice would be r8, r10 and r11.